src/optimization/DAG.cpp

```diff
@@ -0,0 +1,149 @@
+//
+
+#include <log.h>
+#include <assert.h>
+#include <iostream>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <stack>
+#include <utility>
+#include <vector>
+
+#include "DAG.h"
+#include "../InstructionWalker.h"
+#include "log.h"
+
+#include "../ControlFlowGraph.h"
+#include "../DebugGraph.h"
+#include "../InstructionWalker.h"
+#include "../Profiler.h"
+
+namespace vc4c {
+
+/* XXX should be replaced with common routines that generate such instructions? */
+/* reading from mutex I/O register means getting mutex */
+static bool isGetLocInstruction (IL * instr) {
+	for (Value v: instr->getArguments()) {
+		if (v.hasRegister(REG_MUTEX)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* writing to mutex I/O register means release of mutex */
+static bool isReleaseLocInstruction (IL * instr) {
+	if (instr->getOutput().has_value()) {
+		auto output = instr->getOutput().value();
+		return output.hasRegister(REG_MUTEX);
+	}
+
+	return false;
+}
+
+DAG::DAG(BasicBlock &bb) {
+	std::map<Value, IL*> map;
+  auto mutexInstructions = std::shared_ptr<std::vector<IL*>>(new std::vector<IL*>());
+	IL * getMutexInstr = nullptr;
+
+	auto setFlagInstructions = std::shared_ptr<std::vector<IL*>>(new std::vector<IL*>());
+	IL * setFlagInstr = nullptr;
+
+	for (auto it = bb.begin(); !it.isEndOfBlock(); it.nextInBlock()) {
+		auto instr = it.get();
+
+		/* Instruction scheduler is responsible for translation to CombinedOperation.
+		 * Not allowed to translate before!
+		 */
+		assert(! dynamic_cast<vc4c::intermediate::CombinedOperation*>(instr));
+
+		if (dynamic_cast<vc4c::intermediate::BranchLabel *>(instr))
+			continue;
+
+		/* deal with mutex lock */
+		if (isGetLocInstruction(instr)) {
+			assert(getMutexInstr == nullptr);
+			assert(mutexInstructions->empty());
+
+			getMutexInstr = instr;
+			mutexInstructions->push_back(instr);
+			continue;
+		}
+
+		if (isReleaseLocInstruction(instr)) {
+			assert (getMutexInstr != nullptr);
+			assert (! mutexInstructions->empty());
+
+			for (auto depInstr : *mutexInstructions) {
+				graph->getOrCreateNode(depInstr).addNeighbor(&graph->getOrCreateNode(instr), true);
+			}
+
+			getMutexInstr = nullptr;
+			mutexInstructions->clear();
+			continue;
+		}
+
+		if (getMutexInstr != nullptr) {
+			mutexInstructions->push_back(instr);
+			graph->getOrCreateNode(getMutexInstr).addNeighbor(&graph->getOrCreateNode(instr), true);
+		}
+
+		/* deal with set_flags and conditinal executions */
+		if (instr->setFlags == vc4c::SetFlag::SET_FLAGS) {
+			setFlagInstr = instr;
+
+			for (auto depInstr : *setFlagInstructions) {
+				graph->getOrCreateNode(depInstr).addNeighbor(&graph->getOrCreateNode(instr), true);
+			}
+
+			setFlagInstructions->clear();
+		}
+
+		if (instr->hasConditionalExecution()){
+			if (setFlagInstr != nullptr) {
+				/* XXX should be change it to an assertion.
+				 * but, in the basic block of end of function, the case of using flags (but not set in the block) appears.
```

Comment:
What do you mean by this? Can you give an example, of where this occurs?

Reply:
```
kernel void sum_f16 (global float a[], global float b[], global float c[]) {
  int id = get_global_id(0);
  float16 x = vload16(id, b);
  float16 y = vload16(id, c);
  vstore16 (x + y, id, a);
}
```

```
(gdb) p method.dumpInstructions()
[D] Fri Dec 29 11:10:08 2017: Basic block ----
[D] Fri Dec 29 11:10:08 2017: label: %0
[D] Fri Dec 29 11:10:08 2017: Block end ----
[D] Fri Dec 29 11:10:08 2017: Basic block ----
[D] Fri Dec 29 11:10:08 2017: label: %18
[D] Fri Dec 29 11:10:08 2017: i32 %group_id.7 = i32 %group_id_x
[D] Fri Dec 29 11:10:08 2017: i32 %local_info.13 = i32 0 (0)
[D] Fri Dec 29 11:10:08 2017: i32 %local_info.14 = shr i32 %local_sizes, i32 %local_info.13
[D] Fri Dec 29 11:10:08 2017: i8 %immediate.17 = loadi i32 255
[D] Fri Dec 29 11:10:08 2017: i8 %local_size.8 = and i32 %local_info.14, i8 %immediate.17
[D] Fri Dec 29 11:10:08 2017: i32 %global_offset.9 = i32 %global_offset_x
[D] Fri Dec 29 11:10:08 2017: i32 %local_info.16 = shr i32 %local_ids, i32 %local_info.13
[D] Fri Dec 29 11:10:08 2017: i8 %local_id.10 = and i32 %local_info.16, i8 %immediate.17
[D] Fri Dec 29 11:10:08 2017: i32 %global_id.11 = mul24 i32 %group_id.7, i8 %local_size.8
[D] Fri Dec 29 11:10:08 2017: i32 %global_id.12 = add i32 %global_offset.9, i32 %global_id.11
[D] Fri Dec 29 11:10:08 2017: i32 %19 = add i32 %global_id.12, i8 %local_id.10 (global_id)
[D] Fri Dec 29 11:10:08 2017: i32 %21 = shl i32 %19, i32 4 (4)
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %address_offset.1 = mul24 i8 4 (4), register elem_num
[D] Fri Dec 29 11:10:08 2017: i32 %use_with_literal.21 = i32 %21
[D] Fri Dec 29 11:10:08 2017: i32 %index_offset.0 = shl i32 %use_with_literal.21, i32 2 (2)
[D] Fri Dec 29 11:10:08 2017: float* %22 = add float* %16, i32 %index_offset.0
[D] Fri Dec 29 11:10:08 2017: <16 x float>* %25 = float* %22
[D] Fri Dec 29 11:10:08 2017: i8 %immediate.23 = v8adds i32 8 (8), i32 8 (8)
[D] Fri Dec 29 11:10:08 2017: register - = sub register elem_num, i8 %immediate.23 (setf )
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %tmu_address.2 = i8 0 (0) (ifnc )
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %tmu_address.2 = add <16 x float>* %25, <16 x i32> %address_offset.1 (ifn )
[D] Fri Dec 29 11:10:08 2017: register tmu0s = <16 x i32> %tmu_address.2
[D] Fri Dec 29 11:10:08 2017: nop (load_tmu0 )
[D] Fri Dec 29 11:10:08 2017: <16 x float> %26 = register sfu_tmu_in
[D] Fri Dec 29 11:10:08 2017: i32 %use_with_literal.25 = i32 %21
[D] Fri Dec 29 11:10:08 2017: i32 %index_offset.3 = shl i32 %use_with_literal.25, i32 2 (2)
[D] Fri Dec 29 11:10:08 2017: float* %27 = add float* %17, i32 %index_offset.3
[D] Fri Dec 29 11:10:08 2017: <16 x float>* %28 = float* %27
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %address_offset.4 = mul24 i8 4 (4), register elem_num
[D] Fri Dec 29 11:10:08 2017: i8 %immediate.27 = v8adds i32 8 (8), i32 8 (8)
[D] Fri Dec 29 11:10:08 2017: register - = sub register elem_num, i8 %immediate.27 (setf )
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %tmu_address.5 = i8 0 (0) (ifnc )
[D] Fri Dec 29 11:10:08 2017: <16 x i32> %tmu_address.5 = add <16 x float>* %28, <16 x i32> %address_offset.4 (ifn )
[D] Fri Dec 29 11:10:08 2017: register tmu0s = <16 x i32> %tmu_address.5
[D] Fri Dec 29 11:10:08 2017: nop (load_tmu0 )
[D] Fri Dec 29 11:10:08 2017: <16 x float> %29 = register sfu_tmu_in
[D] Fri Dec 29 11:10:08 2017: <16 x float> %30 = fadd <16 x float> %26, <16 x float> %29
[D] Fri Dec 29 11:10:08 2017: i32 %use_with_literal.29 = i32 %21
[D] Fri Dec 29 11:10:08 2017: i32 %index_offset.6 = shl i32 %use_with_literal.29, i32 2 (2)
[D] Fri Dec 29 11:10:08 2017: float* %31 = add float* %15, i32 %index_offset.6
[D] Fri Dec 29 11:10:08 2017: <16 x float>* %32 = float* %31
[D] Fri Dec 29 11:10:08 2017: mutex_acq
[D] Fri Dec 29 11:10:08 2017: register vpw_setup = loadi i32 6656
[D] Fri Dec 29 11:10:08 2017: register vpm = <16 x float> %30
[D] Fri Dec 29 11:10:08 2017: register vpw_setup = loadi i32 2156937216
[D] Fri Dec 29 11:10:08 2017: register vpw_setup = loadi i32 3221291008
[D] Fri Dec 29 11:10:08 2017: register vpw_addr = <16 x float>* %32
[D] Fri Dec 29 11:10:08 2017: register - = register vpw_wait
[D] Fri Dec 29 11:10:08 2017: mutex_rel
[D] Fri Dec 29 11:10:08 2017: Block end ----
[D] Fri Dec 29 11:10:08 2017: Basic block ----
[D] Fri Dec 29 11:10:08 2017: label: %end_of_function
[D] Fri Dec 29 11:10:08 2017: i32 %group_loop_size = register unif
[D] Fri Dec 29 11:10:08 2017: br.ifzc %start_of_function (on i32 %group_loop_size) (ifzc )
[D] Fri Dec 29 11:10:08 2017: Block end ----
```

In block %end_of_function, it occurs.

Reply:
In this case, we can fuse %end_of_function and %18. So This problem can be solved.

Reply:
Is it okay not to set setflag?
https://github.com/doe300/VC4C/blob/master/src/optimization/Combiner.cpp#L634

Reply:
Yes, the setting of flags for conditional branches is inserted (together with the nops for branch-delay) in `extendBranches` in `CodeGenerator.cpp`. Similarily, the start- and stop-segments are added in the CodeGenerator.

I will move them into optimization steps, so you have the final code-layout (including some instructions not mapped to machine code, like labels) to work with.